packages/eslint-plugin/src/rules/avoid-explicit-function-component-type/avoid-explicit-function-component-type.ts

```diff
@@ -0,0 +1,166 @@
+import { basename } from 'node:path';
+
+import {
+  AST_NODE_TYPES,
+  type TSESLint,
+  type TSESTree,
+} from '@typescript-eslint/utils';
+import {
+  isNodeOfType,
+  isNodeOfTypes,
+} from '@typescript-eslint/utils/ast-utils';
+
+import { createRule } from '../../helpers/create-rule';
+
+const isFunctionComponentName = (name: string) => /[A-Z]/.test(name.charAt(0));
+
+const isVariableDeclearatorAFunctionComponent = (
+  declarator: TSESTree.VariableDeclarator,
+): boolean =>
+  isNodeOfType(AST_NODE_TYPES.Identifier)(declarator.id) &&
+  isFunctionComponentName(declarator.id.name) &&
+  isNodeOfTypes([
+    AST_NODE_TYPES.ArrowFunctionExpression,
+    AST_NODE_TYPES.FunctionExpression,
+  ])(declarator.init);
+
+const getTypeNameFromVariableDeclearator = (
+  declarator: TSESTree.VariableDeclarator,
+  context: Readonly<
+    TSESLint.RuleContext<'avoidExplicitFunctionComponentType', []>
+  >,
+): string | undefined => {
+  if (
+    isNodeOfType(AST_NODE_TYPES.TSTypeAnnotation)(
+      declarator.id.typeAnnotation,
+    ) &&
+    isNodeOfType(AST_NODE_TYPES.TSTypeReference)(
+      declarator.id.typeAnnotation.typeAnnotation,
+    )
+  ) {
+    // let a: TypeName ----> TypeName
+    if (
+      isNodeOfType(AST_NODE_TYPES.Identifier)(
+        declarator.id.typeAnnotation.typeAnnotation.typeName,
+      )
+    ) {
+      return declarator.id.typeAnnotation.typeAnnotation.typeName.name;
+    }
+
+    // let a: TypeName<T> ----> TypeName
+    if (
+      isNodeOfType(AST_NODE_TYPES.TSQualifiedName)(
+        declarator.id.typeAnnotation.typeAnnotation.typeName,
+      )
+    ) {
+      return context.sourceCode.getText(
+        declarator.id.typeAnnotation.typeAnnotation.typeName,
+      );
+    }
+  }
+
+  return undefined;
+};
+
+const getTypeArgumentsTextFromVariableDeclearator = (
+  declarator: TSESTree.VariableDeclarator,
+  context: Readonly<
+    TSESLint.RuleContext<'avoidExplicitFunctionComponentType', []>
+  >,
+): string[] | undefined => {
+  if (
+    isNodeOfType(AST_NODE_TYPES.TSTypeAnnotation)(
+      declarator.id.typeAnnotation,
+    ) &&
+    isNodeOfType(AST_NODE_TYPES.TSTypeReference)(
+      declarator.id.typeAnnotation.typeAnnotation,
+    ) &&
+    declarator.id.typeAnnotation.typeAnnotation.typeArguments &&
+    declarator.id.typeAnnotation.typeAnnotation.typeArguments.params.length > 0
+  ) {
+    const typeArgumentsNode =
+      declarator.id.typeAnnotation.typeAnnotation.typeArguments;
+    // let a: TypeName<T1, T2> ----> ["T1", "T2"]
+    return typeArgumentsNode.params.map((param) =>
+      context.sourceCode.getText(param),
+    );
+  }
+
+  return undefined;
+};
+
+export const disallowedTypeNames = [
+  'FC',
```

Comment:
🤔 Maybe `React.Component` type should also be checked.

Reply:
`React.Component` is for class components; it may not be related to this rule. 🤔

Reply:
Sorry, I got it wrong. The type that should be checked is `React.ComponentType`.

Reply:
Regarding the rule applied to variable declarations, it is unusual for someone to write `let MyComponent: React.ComponentType = () => {}`. I think it is not necessary to check this type.

How do you think? @liby

Reply:
> it is unusual for someone to write let MyComponent: React.ComponentType = () => {}

While I agree that this is rare, it does happen occasionally. I believe this falls under the category of a nice-to-have rather than a necessity. Searched on GitHub, Some people write like this.

Regarding the naming of the rule, which is currently `avoid-explicit-function-component-type`, I think it would be beneficial to rename or split this rule into several smaller rules. This approach would facilitate better organization and future use.

For instance, we could consider the following names for the rules:

- `avoid-explicit-function-component-type`
- `avoid-explicit-class-component-type`
- `avoid-explicit-react-component-type` or `avoid-explicit-component-type`